cirq/testing/order_tester.py

```diff
@@ -0,0 +1,230 @@
+equivalence groups or items to it. The order tester will check that
+the items within each group are all equal to each other, and every new
+added item or group is strictly ascending with regard to the previously
+added items or groups.
+
+It will also check that a==b implies hash(a)==hash(b).
+"""
+
+from typing import Any
+
+from cirq.testing.equals_tester import EqualsTester
+
+
+class OrderTester(EqualsTester):
+    """Tests ordering against user-provided disjoint ordered groups or items."""
+
+    def __init__(self):
+        self.groups = [(ClassSmallerThanEverythingElse(),)]
+
+    def _verify_ascending(self, v1, v2):
+        """Verifies that (v1, v2) is a strictly ascending sequence."""
+        lt_1 = NotImplemented
+        not_lt_2 = NotImplemented
+        gt_2 = NotImplemented
+        not_gt_1 = NotImplemented
+        le_1 = NotImplemented
+        not_le_2 = NotImplemented
+        ge_2 = NotImplemented
+        not_ge_1 = NotImplemented
+
+        try:
+            lt_1 = v1 < v2
+        except TypeError:
+            pass
+
+        try:
+            not_lt_2 = not v2 < v1
+        except TypeError:
+            pass
+
+        assert(lt_1, not_lt_2) in [
+            (True, True),
+            (True, NotImplemented),
+            (NotImplemented, True),
+            (NotImplemented, NotImplemented)
+                ]
+
+        try:
+            gt_2 = not hasattr(v2, '__gt__') or v2 > v1
+        except TypeError:
+            pass
+
+        try:
+            not_gt_1 = not hasattr(v1, '__gt__') or not v1 > v2
+        except TypeError:
+            pass
+
+        assert(gt_2, not_gt_1) in [
+            (True, True),
+            (True, NotImplemented),
+            (NotImplemented, True),
+            (NotImplemented, NotImplemented),
+                ]
+
+        # at least one strict ordering operator should work
+        assert(lt_1, gt_2) in [
+            (True, True),
+            (True, NotImplemented),
+            (NotImplemented, True),
+        ]
+
+        try:
+            le_1 = not hasattr(v1, '__le__') or v1 <= v2
+        except TypeError:
+            pass
+
+        try:
+            not_le_2 = not hasattr(v2, '__le__') or not v2 <= v1
+        except TypeError:
+            pass
+
+        assert(le_1, not_le_2) in [
+            (True, True),
+            (True, NotImplemented),
+            (NotImplemented, True),
+            (NotImplemented, NotImplemented),
+                ]
+
+        try:
+            ge_2 = not hasattr(v2, '__ge__') or v2 >= v1
+        except TypeError:
+            pass
+
+        try:
+            not_ge_1 = not hasattr(v1, '__ge__') or not v1 >= v2
+        except TypeError:
+            pass
+
+        assert(ge_2, not_ge_1) in [
+            (True, True),
+            (True, NotImplemented),
+            (NotImplemented, True),
+            (NotImplemented, NotImplemented),
+                ]
+
+    def verify_ascending_group(self, *group_items: Any):
```

Comment:
I don't think this should be a separate available public method. The user should not be choosing whether to add+verify or just verify a group or not. Just always add+verify.

Merge this back into `add_ascending_equivalence_group` or make it private (prefix with `_`).